src/models/PropStore.ts

```diff
@@ -0,0 +1,49 @@
+import { InjectionKey } from 'vue'
+import { CreateRouteOptions, isWithComponent, isWithComponents } from '@/types/createRouteOptions'
+import { ResolvedRoute } from '@/types/resolved'
+import { MaybePromise } from '@/types/utilities'
+
+export const propStoreKey: InjectionKey<PropStore> = Symbol()
+
+type NamedComponentProps = { name: string, props?: (params: Record<string, unknown>) => MaybePromise<Record<string, unknown>> }
+
+export class PropStore {
+  private readonly store = new Map<string, MaybePromise<Record<string, unknown>>>()
+
+  public setProps(route: ResolvedRoute): void {
+    this.store.clear()
+
+    route.matches
+      .flatMap(match => this.getComponentProps(match))
+      .forEach(({ name, props }) => {
+        if (props) {
+          const key = this.getPropKey(route, name)
```

Comment:
I think there's a bug hidden here with unnamed routes. Since `route` here is the resolved route, every key will have the same route id and props. Which means `name` is the only distinction here. So won't any matches that are unnamed share the same key? Which means the props will get overwritten. 

Doesn't `match` have an id as well? I think we want to use that id a not the name. 

Right? Or am I missing something.

Reply:
yeah, this is a good catch